### DISCUSSION TITLE
Configuration to disable Android sensors, and query more sensors as needed.

### DISCUSSION THREAD

**Discussion by Alex2782** [godotengine/godot-proposals#7645] [Platforms]
I could not find any settings to disable the sensors (to save battery and CPU). 
Normal apps and games will rarely use the sensor data.

There are other sensors, e.g. `TYPE_LIGHT`, which many smartphones should have. I think this sensor could also be used creatively in an app or game.

https://developer.android.com/guide/topics/sensors/sensors_overview

> As a best practice you should always disable sensors you don't need, especially when your activity is paused. Failing to do so can drain the battery in just a few hours because some sensors have substantial power requirements and can use up battery power quickly. 

https://github.com/godotengine/godot/blob/master/platform/android/java/lib/src/org/godotengine/godot/Godot.kt#L91
https://github.com/godotengine/godot/blob/master/platform/android/java/lib/src/org/godotengine/godot/Godot.kt#L496

---

**Sensor Rate-Limiting**
https://developer.android.com/guide/topics/sensors/sensors_overview#sensors-rate-limiting

> To protect potentially sensitive information about users, if your app targets Android 12 (API level 31) or higher, the system places a limit on the refresh rate of data from certain motion sensors and position sensors. This data includes values recorded by the device's [accelerometer](https://developer.android.com/guide/topics/sensors/sensors_motion#sensors-motion-accel), [gyroscope](https://developer.android.com/guide/topics/sensors/sensors_motion#sensors-motion-gyro), and [geomagnetic field sensor](https://developer.android.com/guide/topics/sensors/sensors_position#sensors-pos-mag).

> If you call the [registerListener()](https://developer.android.com/reference/android/hardware/SensorManager#registerListener(android.hardware.SensorEventListener,%20android.hardware.Sensor,%20int,%20int)) method to [monitor sensor events](https://developer.android.com/guide/topics/sensors/sensors_overview#sensors-monitor), the sensor sampling rate is limited to 200 Hz. This is true for all overloaded variants of the registerListener() method.

`android.permission.HIGH_SAMPLING_RATE_SENSORS`

Is the limitation still sufficient for XR apps?  
In the 'Android Export' dialog could not find this permission yet.

**Comment by Calinou** [comment]
> Is the limitation still sufficient for XR apps?

200 Hz should be sufficient, as XR displays generally don't exceed 120 Hz.

**Comment by Gromph** [comment]
We have some ANR's reported to Google Play with this call stack:
```
"main" tid=1 Native
  #00  pc 0x000000000007f33c  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #01  pc 0x000000000014b094  /apex/com.android.runtime/lib64/libart.so (art::ConditionVariable::WaitHoldingLocks(art::Thread*)+148)
  #02  pc 0x0000000000590744  /apex/com.android.runtime/lib64/libart.so (art::GoToRunnable(art::Thread*) (.llvm.14603686783979941606)+460)
  #03  pc 0x0000000000590534  /apex/com.android.runtime/lib64/libart.so (art::JniMethodEnd(unsigned int, art::Thread*)+28)
  at android.os.BinderProxy.transactNative (Native method)
  at android.os.BinderProxy.transact (BinderProxy.java:510)
  at android.hardware.display.IDisplayManager$Stub$Proxy.getDisplayInfo (IDisplayManager.java:722)
  at android.hardware.display.DisplayManagerGlobal.getDisplayInfo (DisplayManagerGlobal.java:140)
  at android.view.Display.updateDisplayInfoLocked (Display.java:1126)
  at android.view.Display.getRotation (Display.java:785)
  at org.godotengine.godot.Godot.getRotatedValues (Godot.kt:866)
  at org.godotengine.godot.Godot.onSensorChanged (Godot.kt:896)
  at android.hardware.SystemSensorManager$SensorEventQueue.dispatchSensorEvent (SystemSensorManager.java:837)
  at android.os.MessageQueue.nativePollOnce (Native method)
  at android.os.MessageQueue.next (MessageQueue.java:336)
  at android.os.Looper.loop (Looper.java:174)
  at android.app.ActivityThread.main (ActivityThread.java:7403)
  at java.lang.reflect.Method.invoke (Native method)
  at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run (RuntimeInit.java:502)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:980)
```
The first thing I did was look for a way to disable monitoring sensors, didn't find one,  so just commented out the code.

**Reply by Alex2782** [reply]
better report this issue also here (button "New issue"): https://github.com/godotengine/godot/issues?q=is%3Aissue%20state%3Aopen%20onSensorChanged

**Reply by Gromph** [reply]
Reported: https://github.com/godotengine/godot/issues/99733

**Comment by Alex2782** [comment]
Version 4.4-beta4


![image](https://github.com/user-attachments/assets/5f2d47b4-0e0e-46db-b1af-18cf3dae288d)
